### ISSUE TITLE
Structured dump of metadata / index

### ISSUE THREAD

**Issue by SebAlbert** [restic/restic#4341]
What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

I'd love to get a full dump of the metadata / index in a repository or snapshot, i.e. inodes (files/directories/symlinks) names, hierarchy, mtimes etc., as well as the list of content blob IDs per file. The metadata dump should be in a structured format for further programmatic use and data analysis. As far as I can see, right now I would have to `cat` elements recursively. What I'd like to have is a full dump all at once.

What are you trying to do? What problem would this solve?
---------------------------------------------------------

I would like to build analysis and exploration tools providing hierarchically aggregated and interactively "drillable" (by directory structure) summaries such as (recursive) number of files, sum of size etc. and especially, leveraging the sub-file-level deduplication, automatically generated cross-references to other places (somewhere else in the directory tree or even across snapshots) with high similarity (percentage of matching content blobs).  I would look to e.g. a PostgreSQL database holding the metadata and appropriate indexes to accomplish this, but I need the metadata first.

This would solve my problem of navigating the file organisational mess I've made over the years and will inevitably continue to make.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

Restic lets me rest assured my backups are done, just like ZFS below lets me rest assured that bit-rot won't erase my memories and valuable archives. I also like how with restic I can keep a (mountable!) snapshot of how I had my files organised in directories before a reorganisation at almost no cost.

**Comment by mfschumann** [comment]
Could `restic find / --json` be a starting point for you? This command gives you quite some metadata (path, size, mtime, permissions) for all files across all snapshots as a JSON string.

**Comment by SebAlbert** [comment]
Yes, that's quite a lot already. What's missing here for me is the content blobs and the hierarchy (I would have to reconstruct it from the paths, but it would be better to have "parent" pointers).

**Comment by konidev20** [comment]
@SebAlbert could you give us an example of what you're expecting? I'm trying something similar will see if I can contribute here.

**Comment by wplapper** [comment]
I am in the process of developing a command which dumps the following tables (all in JSON format):
  snapshots: "snapshots": [
                {
                  "snapID": "<snapid 1>",
                  "short": "<short snapid 1>"
                  "snapdata": {
                    normal JSON output for a snapshot 1
                  }
                },
                ...
                {
                  "snapID": "<snapid n>",
                  "short": "<short snapid n>"
                  "snapdata": {
                    normal JSON output for a snapshot n
                  }
                }
              ]

  index: "index": [
          <normal JSON ouput for an index entry 1>,
          ...
          <normal JSON ouput for an index entry n>
          ]

  file: "file_meta_data": [
          {
            "id": "<metablob 1>",
            "path": "<directory path for metablob 1>",
            "nodes": <normal JSON output for a metablob 1>
          },
          ...
          {
            "id": "<metablob n>",
            "path": "<directory path for metablob n>",
            "nodes": <normal JSON output for a metablob n>
          }
        ]

  parent-child: "parent-child-map": [
                  {
                    "parent": "<parent 1 metablob>",
                    "children": [
                      <child 1 metablob>,
                      ...
                      <child n metablob>
                    ]
                  },
                  ...
                  {
                    "parent": "<parent 1 metablob>",
                    "children": [
                      <child 1 metablob>,
                      ...
                      <child n metablob>
                    ]
                  }
                ]

  child-parent: "snapID-child-parent-map": [
                  {
                    "snapid": "<current snapshot>",
                    "child-parent-map": [
                      {
                        "child": "<child 1 metablob>",
                        "parent": "<parent 1 metablob>"
                      },
                      ...
                      {
                        "child": "<child n metablob>",
                        "parent": "<parent n metablob>"
                      }
                    ]
                  }
                ]

  flat-topology: "topology": [
                  {
                    "snapid": "<current snapshot>",
                    "blobs": [
                      "<metablob 1>",
                      ...
                      "<metablob n>"
                    ]
                  },
                ]

So far, I have done nothing specific for data blobs, but they are listed in the "file_meta_data" section. I will attach the actual output of the restic project below.
[restic_cmd_rewrite_include-metadata.txt](https://github.com/user-attachments/files/17285055/restic_cmd_rewrite_include-metadata.txt)
